File: skills/review/scripts/discover.ts
import { spawnSync } from "node:child_process";
import { existsSync, mkdtempSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";

/**
 * Review discovery: detects target, filters files, groups by domain, writes per-group diffs.
 *
 * Outputs JSON to stdout. Called by the review skill orchestrator.
 *
 * Usage:
 *     bun run discover.ts                     # staged changes, fallback to HEAD~1
 *     bun run discover.ts staged              # staged changes only
 *     bun run discover.ts HEAD                # last commit
 *     bun run discover.ts 42                  # PR #42
 *     bun run discover.ts main...HEAD         # branch comparison
 *     bun run discover.ts src/foo.ts src/bar  # specific files
 */

const MAX_GROUPS = 8;

function git(...args: string[]): string {
  const r = spawnSync("git", args, { encoding: "utf-8" });
  return r.status === 0 ? (r.stdout?.trim() ?? "") : "";
}

type Target = {
  description: string;
  cmdParts: string[] | null;
  isGitDiff: boolean;
};

function detectTarget(rawArg: string): Target {
  const arg = rawArg.trim();

  if (!arg || arg === "staged") {
    if (git("diff", "--cached", "--name-only")) {
      return {
        description: "staged changes",
        cmdParts: ["git", "diff", "--cached"],
        isGitDiff: true,
      };
    }
    return {
      description: "HEAD~1 (last commit)",
      cmdParts: ["git", "diff", "HEAD~1"],
      isGitDiff: true,
    };
  }

  if (arg === "HEAD" || arg === "latest") {
    return {
      description: "HEAD~1 (last commit)",
      cmdParts: ["git", "diff", "HEAD~1"],
      isGitDiff: true,
    };
  }

  // PR number or URL — normalize "PR #67", "PR 67", "#67" → "67"
  const prMatch = arg.match(/^(?:PR\s*)?#?(\d+)$/i);
  let prNum: string | null = prMatch ? prMatch[1] : null;
  if (!prNum && arg.includes("github.com") && arg.includes("/pull/")) {
    prNum = arg.replace(/\/$/, "").split("/").pop() ?? null;
  }
  if (prNum) {
    return {
      description: `PR #${prNum}`,
      cmdParts: ["gh", "pr", "diff", prNum],
      isGitDiff: false,
    };
  }

  // Branch ref (contains .. or ...)
  if (arg.includes("...") || arg.includes("..")) {
    return {
      description: arg,
      cmdParts: ["git", "diff", arg],
      isGitDiff: true,
    };
  }

  // File path(s) — check if any exist on disk
  const paths = arg.split(/\s+/);
  const existing = paths.filter((p) => existsSync(p));
  if (existing.length > 0) {
    return {
      description: `files: ${existing.join(", ")}`,
      cmdParts: null,
      isGitDiff: false,
    };
  }

  // Last resort: try as a git ref
  return { description: arg, cmdParts: ["git", "diff", arg], isGitDiff: true };
}

function skipFile(filename: string): boolean {
  const skips = [
    "pnpm-lock.yaml",
    "package-lock.json",
    "yarn.lock",
    "/generated/",
    "generated/",
  ];
  return skips.some((p) => filename.includes(p));
}

function getFilesGit(cmdParts: string[]): {
  files: string[];
  deletedCount: number;
} {
  const r = spawnSync(cmdParts[0], [...cmdParts.slice(1), "--name-status"], {
    encoding: "utf-8",
  });
  if (r.status !== 0 || !r.stdout?.trim()) {
    return { files: [], deletedCount: 0 };
  }

  const files: string[] = [];
  let deleted = 0;
  for (const line of r.stdout.trim().split("\n")) {
    const parts = line.split("\t");
    if (parts.length < 2) continue;
    const status = parts[0];
    if (status === "R100") continue;
    const filename =
      status.startsWith("R") && parts.length > 2 ? parts[2] : parts[1];
    if (skipFile(filename)) continue;
    if (status === "D") {
      deleted++;
      continue;
    }
    files.push(filename);
  }
  return { files, deletedCount: deleted };
}

function getFilesFromDiff(diffText: string): string[] {
  const files: string[] = [];
  for (const line of diffText.split("\n")) {
    if (line.startsWith("diff --git")) {
      const bIdx = line.indexOf(" b/");
      if (bIdx !== -1) {
        const f = line.slice(bIdx + 3);
        if (!skipFile(f)) files.push(f);
      }
    }
  }
  return files;
}

function groupFiles(files: string[]): Record<string, string[]> {
  const groups: Record<string, string[]> = {};

  for (const f of files) {
    const parts = f.split("/");
    let key: string;
    if (parts.length === 1) {
      key = "root config";
    } else if (parts[0] === "src" && parts.length >= 2) {
      key = `src/${parts[1]}`;
    } else if (["test", "tests", "__tests__", "spec"].includes(parts[0])) {
      key = parts.length >= 2 ? `src/${parts[1]}` : "tests";
    } else {
      key = parts[0];
    }
    (groups[key] ??= []).push(f);
  }

  // Split large groups by next level
  const refined: Record<string, string[]> = {};
  for (const [key, flist] of Object.entries(groups)) {
    if (flist.length > 10 && key.includes("/")) {
      const subgroups: Record<string, string[]> = {};
      for (const f of flist) {
        const parts = f.split("/");
        const subkey = parts.length >= 3 ? parts.slice(0, 3).join("/") : key;
        (subgroups[subkey] ??= []).push(f);
      }
      if (Object.keys(subgroups).length > 1) {
        Object.assign(refined, subgroups);
      } else {
        refined[key] = flist;
      }
    } else {
      refined[key] = flist;
    }
  }

  // Merge single-file groups into nearest neighbor
  const final: Record<string, string[]> = {};
  const singles: Record<string, string[]> = {};
  for (const [key, flist] of Object.entries(refined)) {
    if (flist.length <= 1) {
      singles[key] = flist;
    } else {
      final[key] = flist;
    }
  }

  if (Object.keys(singles).length > 0) {
    if (Object.keys(final).length > 0) {
      const target =
        "root config" in final
          ? "root config"
          : Object.keys(final).reduce((a, b) =>
              final[a].length >= final[b].length ? a : b,
            );
      for (const flist of Object.values(singles)) {
        final[target].push(...flist);
      }
    } else {
      final["all files"] = Object.values(singles).flat();
    }
  }

  // Cap at MAX_GROUPS by merging smallest groups
  while (Object.keys(final).length > MAX_GROUPS) {
    const smallestKey = Object.keys(final).reduce((a, b) =>
      final[a].length <= final[b].length ? a : b,
    );
    const smallestFiles = final[smallestKey];
    delete final[smallestKey];

    // Merge into the most related group (longest common prefix) or the largest
    const bestTarget = Object.keys(final).reduce((a, b) => {
      const prefixA = commonPrefixLength(smallestKey, a);
      const prefixB = commonPrefixLength(smallestKey, b);
      if (prefixA !== prefixB) return prefixA > prefixB ? a : b;
      return final[a].length <= final[b].length ? a : b;
    });
    final[bestTarget].push(...smallestFiles);
  }

  return final;
}

function commonPrefixLength(a: string, b: string): number {
  let i = 0;
  while (i < a.length && i < b.length && a[i] === b[i]) i++;
  return i;
}

function filterDiffForFiles(rawDiff: string, targetFiles: string[]): string {
  const targetSet = new Set(targetFiles);
  const result: string[] = [];
  let currentFile: string | null = null;
  let currentLines: string[] = [];

  for (const line of rawDiff.split("\n")) {
    if (line.startsWith("diff --git")) {
      if (currentFile && targetSet.has(currentFile)) {
        result.push(...currentLines);
      }
      currentLines = [line];
      const bIdx = line.indexOf(" b/");
      currentFile = bIdx !== -1 ? line.slice(bIdx + 3) : null;
    } else {
      currentLines.push(line);
    }
  }

  if (currentFile && targetSet.has(currentFile)) {
    result.push(...currentLines);
  }
  return result.join("\n");
}

function main(): void {
  const rawArg = process.argv.slice(2).join(" ");
  const { description, cmdParts, isGitDiff } = detectTarget(rawArg);

  // Direct file review (no diff)
  if (cmdParts === null) {
    const filePaths = rawArg.split(/\s+/);
    const existing = filePaths.filter((p) => existsSync(p));
    if (existing.length === 0) {
      console.log(JSON.stringify({ error: "no_changes" }));
      return;
    }
    const tmpDir = mkdtempSync(join(tmpdir(), "review-"));
    const diffFile = join(tmpDir, "group-1.diff");
    const parts: string[] = [];
    for (const p of existing) {
      parts.push(`=== ${p} ===\n`);
      parts.push(readFileSync(p, "utf-8"));
      parts.push("\n");
    }
    writeFileSync(diffFile, parts.join(""));
    console.log(
      JSON.stringify({
        target: description,
        total_files: existing.length,
        deleted_files: 0,
        groups: [
          {
            name: "all files",
            files: existing,
            file_count: existing.length,
            diff_file: diffFile,
          },
        ],
      }),
    );
    return;
  }

  // Get file list
  let rawDiff: string | null = null;
  let deletedCount = 0;
  let files: string[];

  if (isGitDiff) {
    const result = getFilesGit(cmdParts);
    files = result.files;
    deletedCount = result.deletedCount;
  } else {
    const r = spawnSync(cmdParts[0], cmdParts.slice(1), {
      encoding: "utf-8",
    });
    if (r.status !== 0 || !r.stdout?.trim()) {
      if (r.status !== 0 && r.stderr?.trim()) {
        console.log(
          JSON.stringify({
            error: "command_failed",
            message: r.stderr.trim(),
          }),
        );
      } else {
        console.log(JSON.stringify({ error: "no_changes" }));
      }
      return;
    }
    rawDiff = r.stdout;
    files = getFilesFromDiff(rawDiff);
  }

  if (files.length === 0) {
    if (deletedCount > 0) {
      console.log(
        JSON.stringify({
          target: description,
          total_files: 0,
          deleted_files: deletedCount,
          groups: [],
          note: `${deletedCount} files were deleted (excluded from review).`,
        }),
      );
    } else {
      console.log(JSON.stringify({ error: "no_changes" }));
    }
    return;
  }

  const tmpDir = mkdtempSync(join(tmpdir(), "review-"));

  // Single group (<=5 files)
  if (files.length <= 5) {
    const diffFile = join(tmpDir, "group-1.diff");
    let diffContent: string;
    if (isGitDiff) {
      const r = spawnSync(cmdParts[0], [...cmdParts.slice(1), "--", ...files], {
        encoding: "utf-8",
      });
      diffContent = r.stdout ?? "";
    } else {
      diffContent = rawDiff!;
    }
    writeFileSync(diffFile, diffContent);
    console.log(
      JSON.stringify({
        target: description,
        total_files: files.length,
        deleted_files: deletedCount,
        groups: [
          {
            name: "all files",
            files,
            file_count: files.length,
            diff_file: diffFile,
          },
        ],
      }),
    );
    return;
  }

  // Multi-group (>5 files)
  const grouped = groupFiles(files);
  const groups: Array<{
    name: string;
    files: string[];
    file_count: number;
    diff_file: string;
  }> = [];

  const sortedEntries = Object.entries(grouped).sort(([a], [b]) =>
    a.localeCompare(b),
  );

  for (let i = 0; i < sortedEntries.length; i++) {
    const [name, flist] = sortedEntries[i];
    const diffFile = join(tmpDir, `group-${i + 1}.diff`);
    let diffContent: string;
    if (isGitDiff) {
      const r = spawnSync(cmdParts[0], [...cmdParts.slice(1), "--", ...flist], {
        encoding: "utf-8",
      });
      diffContent = r.stdout ?? "";
    } else {
      diffContent = filterDiffForFiles(rawDiff!, flist);
    }
    writeFileSync(diffFile, diffContent);
    groups.push({
      name,
      files: flist,
      file_count: flist.length,
      diff_file: diffFile,
    });
  }

  console.log(
    JSON.stringify({
      target: description,
      total_files: files.length,
      deleted_files: deletedCount,
      groups,
    }),
  );
}

main();

File: skills/review/scripts/discover.py
#!/usr/bin/env python3
"""
Review discovery: detects target, filters files, groups by domain, writes per-group diffs.

Outputs JSON to stdout. Called by the review skill orchestrator.

Usage:
    python3 discover.py                     # staged changes, fallback to HEAD~1
    python3 discover.py staged              # staged changes only
    python3 discover.py HEAD                # last commit
    python3 discover.py 42                  # PR #42
    python3 discover.py main...HEAD         # branch comparison
    python3 discover.py src/foo.ts src/bar  # specific files
"""

import json
import os
import re
import subprocess
import sys
import tempfile
from collections import defaultdict
from pathlib import Path

MAX_GROUPS = 8


def git(*args):
    r = subprocess.run(["git", *args], capture_output=True, text=True)
    return r.stdout.strip() if r.returncode == 0 else ""


def detect_target(raw_arg):
    """
    Returns (description, diff_base_cmd, is_git_diff).
    - diff_base_cmd: list of command parts for generating diffs
    - is_git_diff: True if we can append --name-status / -- files
    """
    arg = raw_arg.strip()

    if not arg or arg == "staged":
        if git("diff", "--cached", "--name-only"):
            return "staged changes", ["git", "diff", "--cached"], True
        return "HEAD~1 (last commit)", ["git", "diff", "HEAD~1"], True

    if arg in ("HEAD", "latest"):
        return "HEAD~1 (last commit)", ["git", "diff", "HEAD~1"], True

    # PR number or URL — normalize "PR #67", "PR 67", "#67" → "67"
    pr_num = None
    pr_match = re.match(r'^(?:PR\s*)?#?(\d+)$', arg, re.IGNORECASE)
    if pr_match:
        pr_num = pr_match.group(1)
    elif "github.com" in arg and "/pull/" in arg:
        pr_num = arg.rstrip("/").split("/")[-1]
    if pr_num:
        return f"PR #{pr_num}", ["gh", "pr", "diff", pr_num], False

    # Branch ref (contains .. or ...)
    if "..." in arg or ".." in arg:
        return arg, ["git", "diff", arg], True

    # File path(s) — check if any exist on disk
    paths = arg.split()
    existing = [p for p in paths if Path(p).exists()]
    if existing:
        return f"files: {', '.join(existing)}", None, False

    # Last resort: try as a git ref
    return arg, ["git", "diff", arg], True


def get_files_git(cmd_parts):
    """Get changed files via git diff --name-status. Returns (reviewable, deleted_count)."""
    r = subprocess.run(cmd_parts + ["--name-status"], capture_output=True, text=True)
    if r.returncode != 0 or not r.stdout.strip():
        return [], 0

    files = []
    deleted = 0
    for line in r.stdout.strip().split("\n"):
        parts = line.split("\t")
        if len(parts) < 2:
            continue
        status = parts[0]
        if status == "R100":
            continue
        filename = parts[2] if status.startswith("R") and len(parts) > 2 else parts[1]
        if _skip_file(filename):
            continue
        if status == "D":
            deleted += 1
            continue
        files.append(filename)
    return files, deleted


def get_files_from_diff(diff_text):
    """Parse file names from raw unified diff output."""
    files = []
    for line in diff_text.split("\n"):
        if line.startswith("diff --git"):
            parts = line.split(" b/", 1)
            if len(parts) == 2:
                f = parts[1]
                if not _skip_file(f):
                    files.append(f)
    return files


def _skip_file(filename):
    """True if the file should be excluded from review."""
    skips = [
        "pnpm-lock.yaml", "package-lock.json", "yarn.lock",
        "/generated/", "generated/",
    ]
    return any(p in filename for p in skips)


def group_files(files):
    """
    Group files by directory. Rules:
    - Root files -> "root config"
    - src/X/... -> "src/X"
    - test dirs -> merge with matching src group
    - Groups >10 files -> split by next directory level
    - Groups with 1 file -> merge into nearest group
    - Cap at MAX_GROUPS by merging smallest groups
    """
    groups = defaultdict(list)
    for f in files:
        parts = f.split("/")
        if len(parts) == 1:
            groups["root config"].append(f)
        elif parts[0] == "src" and len(parts) >= 2:
            groups[f"src/{parts[1]}"].append(f)
        elif parts[0] in ("test", "tests", "__tests__", "spec"):
            src_key = f"src/{parts[1]}" if len(parts) >= 2 else "tests"
            groups[src_key].append(f)
        else:
            groups[parts[0]].append(f)

    # Split large groups by next level
    refined = {}
    for key, flist in groups.items():
        if len(flist) > 10 and "/" in key:
            subgroups = defaultdict(list)
            for f in flist:
                parts = f.split("/")
                subkey = "/".join(parts[:3]) if len(parts) >= 3 else key
                subgroups[subkey].append(f)
            if len(subgroups) > 1:
                refined.update(subgroups)
            else:
                refined[key] = flist
        else:
            refined[key] = flist

    # Merge single-file groups into nearest neighbor
    final = {}
    singles = {}
    for key, flist in refined.items():
        if len(flist) <= 1:
            singles[key] = flist
        else:
            final[key] = flist

    if singles:
        if final:
            target = "root config" if "root config" in final else max(final, key=lambda k: len(final[k]))
            for flist in singles.values():
                final[target].extend(flist)
        else:
            final["all files"] = [f for flist in singles.values() for f in flist]

    # Cap at MAX_GROUPS by merging smallest groups
    while len(final) > MAX_GROUPS:
        smallest_key = min(final, key=lambda k: len(final[k]))
        smallest_files = final.pop(smallest_key)
        # Merge into the most related group (longest common prefix) or the largest
        best_target = max(final, key=lambda k: (
            len(os.path.commonprefix([smallest_key, k])),
            -len(final[k]),
        ))
        final[best_target].extend(smallest_files)

    return final


def filter_diff_for_files(raw_diff, target_files):
    """Extract sections of a unified diff matching target files."""
    target_set = set(target_files)
    result = []
    current_file = None
    current_lines = []

    for line in raw_diff.split("\n"):
        if line.startswith("diff --git"):
            if current_file in target_set:
                result.extend(current_lines)
            current_lines = [line]
            parts = line.split(" b/", 1)
            current_file = parts[1] if len(parts) == 2 else None
        else:
            current_lines.append(line)

    if current_file in target_set:
        result.extend(current_lines)
    return "\n".join(result)


def main():
    raw_arg = " ".join(sys.argv[1:])
    desc, cmd_parts, is_git = detect_target(raw_arg)

    # Direct file review (no diff)
    if cmd_parts is None:
        file_paths = raw_arg.split()
        existing = [p for p in file_paths if Path(p).exists()]
        if not existing:
            print(json.dumps({"error": "no_changes"}))
            return
        tmpdir = tempfile.mkdtemp(prefix="review-")
        diff_file = os.path.join(tmpdir, "group-1.diff")
        parts = []
        for p in existing:
            parts.append(f"=== {p} ===\n")
            with open(p) as f:
                parts.append(f.read())
            parts.append("\n")
        with open(diff_file, "w") as f:
            f.write("".join(parts))
        print(json.dumps({
            "target": desc,
            "total_files": len(existing),
            "deleted_files": 0,
            "groups": [{"name": "all files", "files": existing,
                        "file_count": len(existing), "diff_file": diff_file}],
        }))
        return

    # Get file list
    raw_diff = None
    deleted_count = 0
    if is_git:
        files, deleted_count = get_files_git(cmd_parts)
    else:
        r = subprocess.run(cmd_parts, capture_output=True, text=True)
        if r.returncode != 0 or not r.stdout.strip():
            if r.returncode != 0 and r.stderr.strip():
                print(json.dumps({"error": "command_failed", "message": r.stderr.strip()}))
            else:
                print(json.dumps({"error": "no_changes"}))
            return
        raw_diff = r.stdout
        files = get_files_from_diff(raw_diff)

    if not files:
        if deleted_count > 0:
            print(json.dumps({
                "target": desc, "total_files": 0, "deleted_files": deleted_count,
                "groups": [],
                "note": f"{deleted_count} files were deleted (excluded from review).",
            }))
        else:
            print(json.dumps({"error": "no_changes"}))
        return

    tmpdir = tempfile.mkdtemp(prefix="review-")

    # Single group (<=5 files)
    if len(files) <= 5:
        diff_file = os.path.join(tmpdir, "group-1.diff")
        if is_git:
            r = subprocess.run(cmd_parts + ["--"] + files, capture_output=True, text=True)
            diff_content = r.stdout
        else:
            diff_content = raw_diff
        with open(diff_file, "w") as f:
            f.write(diff_content)
        print(json.dumps({
            "target": desc,
            "total_files": len(files),
            "deleted_files": deleted_count,
            "groups": [{"name": "all files", "files": files,
                        "file_count": len(files), "diff_file": diff_file}],
        }))
        return

    # Multi-group (>5 files)
    grouped = group_files(files)
    groups = []
    for i, (name, flist) in enumerate(sorted(grouped.items()), 1):
        diff_file = os.path.join(tmpdir, f"group-{i}.diff")
        if is_git:
            r = subprocess.run(cmd_parts + ["--"] + flist, capture_output=True, text=True)
            diff_content = r.stdout
        else:
            diff_content = filter_diff_for_files(raw_diff, flist)
        with open(diff_file, "w") as f:
            f.write(diff_content)
        groups.append({
            "name": name, "files": flist,
            "file_count": len(flist), "diff_file": diff_file,
        })

    print(json.dumps({
        "target": desc,
        "total_files": len(files),
        "deleted_files": deleted_count,
        "groups": groups,
    }))


if __name__ == "__main__":
    main()
